File: shors.rb
class Shors
    def find_gcd(number_1, number_2)
        return number_1.gcd(number_2)
    end

    def shors_algorithm(number_to_factor)
        if !number_to_factor.is_a? Integer
            return
        end
        N = number_to_factor
        puts "The number is: #{N}"
        K = rand(1..number_to_factor)
        puts "The factor number is #{K}"
        GCD = find_gcd(N, K)
        puts "The GCD is #{GCD}"
        if GCD != 1
            puts "#{GCD} is a factor of #{N}"
        else
            transformations = Array.new()
            q = 1
            iterations = 0
            while true
                q = (q*K) % N
                transformations.push(q)
                iterations += 1
                puts "This iteration #{iterations}, remainder #{q}, N #{N}, and factor #{K}"
                if q == 1
                    break
                end
            end
            if iterations % 2 != 0
                puts "Number of iterations is not even"
                shors_algorithm(N)
            else
                p = transformations[int((iterations/2)-1)]
                puts "This is p #{p}"
                if p + 1 == N
                    puts "P + 1 == N, recalling function"
                    shors_algorithm(N)
                else
                    puts "This is the #{int(iterations/2)} transformation"
                    puts "#{find_gcd(p+1, N)} and #{find_gcd(p-1, N)} are factors of #{N}"
                end
            end
        end
    end
end

shors = Shors.new()
shors.shors_algorithm(gets.chomp.to_i)

File: shors.java
import java.math.BigInteger;
import java.util.*;

class shors {
    public static BigInteger find_gcd(BigInteger number_1, BigInteger number_2) {
        return number_1.gcd(number_2);
    }

    public static int get_random_number(int lower_bound, int upper_bound){
        return (int) ((Math.random() * (upper_bound - lower_bound)) + lower_bound);
    }

    public static void shors_algorithm(int number_to_factor){
        int N = number_to_factor;
        int K = get_random_number(1, number_to_factor);
        BigInteger gcd = find_gcd(BigInteger.valueOf(N), BigInteger.valueOf(K));
        int GCD = gcd.intValue();
        if (GCD != 1)
            System.out.println(GCD + " is a factor of " + N);
        else{
            List<Integer> transformations = new ArrayList();
            int q = 1;
            int iterations = 0;
            while (true){
                q = (q*K) % N;
                transformations.add(q);
                iterations++;
                if (q == 1)
                    break;
            }
            if (iterations % 2 != 0){
                System.out.println("Number of iterations is not even");
                shors_algorithm(N);
            }
            else{
                int p = transformations.get(((iterations/2)-1));
                if (p + 1 == N){
                    System.out.println("P=1 == N, recalling function");
                    shors_algorithm(N);
                }
                else{
                    System.out.println("This is the " + iterations/2 + " transformation");
                    System.out.println(find_gcd(BigInteger.valueOf(p+1), BigInteger.valueOf(N)) + " and " + find_gcd(BigInteger.valueOf(p-1), BigInteger.valueOf(N)) + " are factors of " + N);
                }
            }
        }
    }

    public static void main(String[] args) {
        shors_algorithm(get_random_number(Integer.MIN_VALUE, Integer.MAX_VALUE));
    }
}